core/controllers/editor_test.py

```diff
@@ -1381,6 +1391,49 @@ def test_for_assign_role_for_exploration(self):
         self.assert_cannot_voiceover(exp_id)
         self.logout()
 
+        # Check that COLLABORATOR 4 who cannot edit but can voiceover,
+        # cannot add a new State 6.
+        self.login(self.COLLABORATOR4_EMAIL)
+        csrf_token = self.get_new_csrf_token()
+
+        # Since, check edit_activity_always return false, therefore
+        # editor will behave as voiceover artist.
+        get_voiceover_swap = self.swap_to_always_return(
+            rights_manager, 'check_can_edit_activity', value=False)
+
+        with get_voiceover_swap:
+            self.assert_cannot_edit(exp_id)
+            self.assert_can_voiceover(exp_id)
+
+            response_dict = self.put_json(
+                '%s/%s' % (feconf.EXPLORATION_DATA_PREFIX, exp_id),
+                {
+                    'version': exploration.version,
+                    'commit_message': 'Added State 4',
+                    'change_list': [{
+                        'cmd': 'add_state',
+                        'state_name': 'State 4'
+                    }]
+                },
+                csrf_token=csrf_token,
+                expected_status_int=500
+            )
```

Comment:
Can we check that the state didn't get added? I know you're checking that the status code is 500, but what if there's a mistake in the backend code where even though the status code is 500, the change is accepted?

Reply:
@U8NWXD Sorry for the late reply. Actually, it took me a lot of time in understanding the codebase and figure out a solution to test the same. I implemented a working solution to the best of my knowledge. Kindly, review it.